symcache/src/heuristics.rs

```diff
@@ -4,65 +4,113 @@ const SIGILL: u32 = 4;
     }
 
-    /// Give the information in the instruction info this returns the
-    /// most accurate instruction.
-    pub fn find_best_instruction(&self) -> u64 {
-        let mut prev = false;
+    /// Returns whether the application attempted to jump to an invalid,
+    /// privileged or misaligned address. This indicates, that certain
+    /// adjustments should be made on the caller instruction address.
+    pub fn is_crash_signal(&self) -> bool {
+        match self.signal {
+            Some(SIGILL) | Some(SIGBUS) | Some(SIGSEGV) => true,
+            _ => false,
+        }
+    }
 
+    /// Determines whether the given address should be adjusted to resolve the
+    /// call site of a stack frame.
+    ///
+    /// This generally applies to all frames except the crashing / suspended
+    /// frame. However, if the process crashed with an illegal instruction,
+    /// even the top-most frame needs to be adjusted to account for the signal
+    /// handler.
+    pub fn should_adjust_caller(&self) -> bool {
+        // All frames other than the crashing frame (or suspended frame for
+        // other threads) report the return address. This address (generally)
+        // points to the instruction after the function call. Therefore, we
+        // need to adjust the caller address for these frames.
         if !self.crashing_frame {
-            prev = true;
-        } else if let Some(ip) = self.ip_reg {
+            return true;
+        }
+
+        // The crashing frame usually contains the actual register contents,
+        // which points to the exact instruction that crashed and must not be
+        // adjusted. A notable exception to this is if we crashed with one of
+        // the crash signals.
+        // TODO: Document reason of this
+        if let Some(ip) = self.ip_reg {
             if ip != self.addr && self.is_crash_signal() {
-                prev = true;
+                return true;
             }
         }
 
-        round_to_instruction_end(if prev {
-            self.get_previous_instruction().unwrap_or(self.addr.saturating_sub(1))
-        } else {
-            self.addr
-        }, self.arch)
+        return false;
     }
-}
 
-fn round_to_instruction_end(addr: u64, arch: Arch) -> u64 {
-    match arch.cpu_family() {
-        CpuFamily::Arm64 => addr - (addr % 4) + 3,
-        CpuFamily::Arm32 => addr - (addr % 2) + 1,
-        _ => addr
+    /// Give the information in the instruction info this returns the
+    /// most accurate instruction.
+    pub fn caller_address(&self) -> u64 {
+        let addr = if self.should_adjust_caller() {
+            self.previous_address()
+        } else {
+            self.aligned_address()
+        };
+
+        // For some crashes on ARM, the aligned address resolved to the wrong
+        // symbol.
+        // TODO: Document details for this
+        match self.arch.cpu_family() {
+            CpuFamily::Arm32 => addr + 1,
+            CpuFamily::Arm64 => addr + 3,
```

Comment:
This is the second one. Do we have an example where this shows effect? Judging by how our symbol lookup works, this **should** make no difference, so I suspect a bug in a different place.